DynamoDB Mock does not handle UpdateExpression implementing an Atomic Counter correctly

Amazon's DynamoDB can do an atomic increment of a counter attribute using an UpdateExpression. See https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithItems.html#WorkingWithItems.AtomicCounters

## Test Case to Reproduce
This tries to run the same test using real boto and a moto mock. Requires valid AWS credentials for the test against the real boto.

```
import unittest
import boto3
import time

from moto import mock_dynamodb2

dynamodb = boto3.resource('dynamodb')


class MotoBugTestCase(unittest.TestCase):
    def setUp(self):
        self.table_name = 'test_moto_bug_ppeach'

        self.primary_key = 'a_key'

    def setup_table(self):
        print('Creating the table: {}'.format(self.table_name))
        table = dynamodb.create_table(
            TableName=self.table_name,
            KeySchema=[{
                'AttributeName': 'primary_key',
                'KeyType': 'HASH'
            }],
            AttributeDefinitions=[{
                'AttributeName': 'primary_key',
                'AttributeType': 'S'
            }],
            ProvisionedThroughput={
                'ReadCapacityUnits': 1,
                'WriteCapacityUnits': 1
            })
        return table

    def teardown_table(self, table):
        print('Deleting the table: {}'.format(self.table_name))
        table.delete()

    def increment_item_and_verify(self):
        table = self.setup_table()
        time.sleep(10)
        item = {'primary_key': self.primary_key, 'num_counter': 0}
        try:
            print('Create item {}'.format(self.primary_key))
            table.put_item(Item=item)

            incr = 2
            print('Incrementing item {} by {}'.format(self.primary_key, incr))
            table.update_item(
                Key={'primary_key': self.primary_key},
                UpdateExpression='SET num_counter = num_counter + :incr',
                ExpressionAttributeValues={':incr': incr})

            print('Fetching item {}'.format(self.primary_key))
            response = table.get_item(Key={'primary_key': self.primary_key})
            print('For item {} got {}'.format(self.primary_key, response['Item']))
            self.assertEqual(2, response['Item']['num_counter'])
        except:
            import traceback
            print(traceback.format_exc())
            raise
        finally:
            self.teardown_table(table)

    def test_real_scenario(self):
        self.increment_item_and_verify()

    @mock_dynamodb2
    def test_moto_scenario(self):
        self.increment_item_and_verify()

```

## Expected Result
Both boto and moto should act the same and both tests should pass.

## Actual Result
The moto test fails because the `num_counter` attribute is set to the string `'num_counter+:incr'` instead of being parsed into an increment operation.

```
$ pipenv run python -m unittest discover -v -b
test_moto_scenario (test.test_moto_bug.MotoBugTestCase) ... FAIL

Stdout:
Creating the table: test_moto_bug_ppeach
Create item a_key
Incrementing item a_key by 2
Fetching item a_key
For item a_key got {u'num_counter': u'num_counter+:incr', u'primary_key': u'a_key'}
Traceback (most recent call last):
  File "/home/ppeach/git_repos/onep/onep-event/test/test_moto_bug.py", line 56, in increment_item_and_verify
    self.assertEqual(2, response['Item']['num_counter'])
  File "/usr/lib/python2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python2.7/unittest/case.py", line 506, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 2 != u'num_counter+:incr'

Deleting the table: test_moto_bug_ppeach
test_real_scenario (test.test_moto_bug.MotoBugTestCase) ... ok

======================================================================
FAIL: test_moto_scenario (test.test_moto_bug.MotoBugTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ppeach/.local/share/virtualenvs/onep-event-JWHCAMFK/local/lib/python2.7/site-packages/moto/core/models.py", line 71, in wrapper
    result = func(*args, **kwargs)
  File "/home/ppeach/git_repos/onep/onep-event/test/test_moto_bug.py", line 69, in test_moto_scenario
    self.increment_item_and_verify()
  File "/home/ppeach/git_repos/onep/onep-event/test/test_moto_bug.py", line 56, in increment_item_and_verify
    self.assertEqual(2, response['Item']['num_counter'])
AssertionError: 2 != u'num_counter+:incr'

Stdout:
Creating the table: test_moto_bug_ppeach
Create item a_key
Incrementing item a_key by 2
Fetching item a_key
For item a_key got {u'num_counter': u'num_counter+:incr', u'primary_key': u'a_key'}
Traceback (most recent call last):
  File "/home/ppeach/git_repos/onep/onep-event/test/test_moto_bug.py", line 56, in increment_item_and_verify
    self.assertEqual(2, response['Item']['num_counter'])
  File "/usr/lib/python2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python2.7/unittest/case.py", line 506, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 2 != u'num_counter+:incr'

Deleting the table: test_moto_bug_ppeach

----------------------------------------------------------------------
Ran 2 tests in 20.340s

FAILED (failures=1)
```

## Environment
moto 1.3.4
boto3 1.7.75
botocore 1.10.75

```
$ pipenv graph
behave==1.2.6
  - enum34 [required: Any, installed: 1.1.6]
  - parse [required: >=1.8.2, installed: 1.8.4]
  - parse-type [required: >=0.4.2, installed: 0.4.2]
    - enum34 [required: Any, installed: 1.1.6]
    - parse [required: >=1.8, installed: 1.8.4]
    - six [required: >=1.11, installed: 1.11.0]
  - six [required: >=1.11, installed: 1.11.0]
  - traceback2 [required: Any, installed: 1.4.0]
    - linecache2 [required: Any, installed: 1.0.0]
confluence-publisher==1.2.1
  - argparse [required: >=1.2.1, installed: 1.1]
  - PyYAML [required: >=3.11, installed: 3.13]
  - requests [required: >=2.4.3, installed: 2.19.1]
    - certifi [required: >=2017.4.17, installed: 2018.8.13]
    - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
    - idna [required: >=2.5,<2.8, installed: 2.7]
    - urllib3 [required: >=1.21.1,<1.24, installed: 1.23]
coverage==4.5.1
grequests==0.3.0
  - gevent [required: Any, installed: 1.3.5]
    - greenlet [required: >=0.4.13, installed: 0.4.14]
  - requests [required: Any, installed: 2.19.1]
    - certifi [required: >=2017.4.17, installed: 2018.8.13]
    - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
    - idna [required: >=2.5,<2.8, installed: 2.7]
    - urllib3 [required: >=1.21.1,<1.24, installed: 1.23]
httpretty==0.9.4
  - six [required: Any, installed: 1.11.0]
jsonmerge==1.5.1
  - jsonschema [required: Any, installed: 2.6.0]
    - functools32 [required: Any, installed: 3.2.3.post2]
moto==1.3.4
  - aws-xray-sdk [required: >=0.93,<0.96, installed: 0.95]
    - jsonpickle [required: Any, installed: 0.9.6]
    - requests [required: Any, installed: 2.19.1]
      - certifi [required: >=2017.4.17, installed: 2018.8.13]
      - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
      - idna [required: >=2.5,<2.8, installed: 2.7]
      - urllib3 [required: >=1.21.1,<1.24, installed: 1.23]
    - wrapt [required: Any, installed: 1.10.11]
  - backports.tempfile [required: Any, installed: 1.0]
    - backports.weakref [required: Any, installed: 1.0.post1]
  - boto [required: >=2.36.0, installed: 2.49.0]
  - boto3 [required: >=1.6.16, installed: 1.7.75]
    - botocore [required: >=1.10.75,<1.11.0, installed: 1.10.75]
      - docutils [required: >=0.10, installed: 0.14]
      - jmespath [required: >=0.7.1,<1.0.0, installed: 0.9.3]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.7.3]
        - six [required: >=1.5, installed: 1.11.0]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.9.3]
    - s3transfer [required: >=0.1.10,<0.2.0, installed: 0.1.13]
      - botocore [required: >=1.3.0,<2.0.0, installed: 1.10.75]
        - docutils [required: >=0.10, installed: 0.14]
        - jmespath [required: >=0.7.1,<1.0.0, installed: 0.9.3]
        - python-dateutil [required: >=2.1,<3.0.0, installed: 2.7.3]
          - six [required: >=1.5, installed: 1.11.0]
      - futures [required: >=2.2.0,<4.0.0, installed: 3.2.0]
  - botocore [required: >=1.9.16, installed: 1.10.75]
    - docutils [required: >=0.10, installed: 0.14]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.9.3]
    - python-dateutil [required: >=2.1,<3.0.0, installed: 2.7.3]
      - six [required: >=1.5, installed: 1.11.0]
  - cookies [required: Any, installed: 2.2.1]
  - cryptography [required: >=2.0.0, installed: 2.3]
    - asn1crypto [required: >=0.21.0, installed: 0.24.0]
    - cffi [required: >=1.7,!=1.11.3, installed: 1.11.5]
      - pycparser [required: Any, installed: 2.18]
    - enum34 [required: Any, installed: 1.1.6]
    - idna [required: >=2.1, installed: 2.7]
    - ipaddress [required: Any, installed: 1.0.22]
    - six [required: >=1.4.1, installed: 1.11.0]
  - docker [required: >=2.5.1, installed: 3.5.0]
    - backports.ssl-match-hostname [required: >=3.5, installed: 3.5.0.1]
    - docker-pycreds [required: >=0.3.0, installed: 0.3.0]
      - six [required: >=1.4.0, installed: 1.11.0]
    - ipaddress [required: >=1.0.16, installed: 1.0.22]
    - requests [required: >=2.14.2,!=2.18.0, installed: 2.19.1]
      - certifi [required: >=2017.4.17, installed: 2018.8.13]
      - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
      - idna [required: >=2.5,<2.8, installed: 2.7]
      - urllib3 [required: >=1.21.1,<1.24, installed: 1.23]
    - six [required: >=1.4.0, installed: 1.11.0]
    - websocket-client [required: >=0.32.0, installed: 0.48.0]
      - six [required: Any, installed: 1.11.0]
  - Jinja2 [required: >=2.7.3, installed: 2.10]
    - MarkupSafe [required: >=0.23, installed: 1.0]
  - jsondiff [required: ==1.1.1, installed: 1.1.1]
  - mock [required: Any, installed: 2.0.0]
    - funcsigs [required: >=1, installed: 1.0.2]
    - pbr [required: >=0.11, installed: 4.2.0]
    - six [required: >=1.9, installed: 1.11.0]
  - pyaml [required: Any, installed: 17.12.1]
    - PyYAML [required: Any, installed: 3.13]
  - python-dateutil [required: >=2.1,<3.0.0, installed: 2.7.3]
    - six [required: >=1.5, installed: 1.11.0]
  - python-jose [required: <3.0.0, installed: 2.0.2]
    - ecdsa [required: <1.0, installed: 0.13]
    - future [required: <1.0, installed: 0.16.0]
    - pycryptodome [required: >=3.3.1,<4.0.0, installed: 3.6.5]
    - six [required: <2.0, installed: 1.11.0]
  - pytz [required: Any, installed: 2018.5]
  - requests [required: >=2.5, installed: 2.19.1]
    - certifi [required: >=2017.4.17, installed: 2018.8.13]
    - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
    - idna [required: >=2.5,<2.8, installed: 2.7]
    - urllib3 [required: >=1.21.1,<1.24, installed: 1.23]
  - responses [required: >=0.9.0, installed: 0.9.0]
    - cookies [required: Any, installed: 2.2.1]
    - mock [required: Any, installed: 2.0.0]
      - funcsigs [required: >=1, installed: 1.0.2]
      - pbr [required: >=0.11, installed: 4.2.0]
      - six [required: >=1.9, installed: 1.11.0]
    - requests [required: >=2.0, installed: 2.19.1]
      - certifi [required: >=2017.4.17, installed: 2018.8.13]
      - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
      - idna [required: >=2.5,<2.8, installed: 2.7]
      - urllib3 [required: >=1.21.1,<1.24, installed: 1.23]
    - six [required: Any, installed: 1.11.0]
  - six [required: >1.9, installed: 1.11.0]
  - werkzeug [required: Any, installed: 0.14.1]
  - xmltodict [required: Any, installed: 0.11.0]
ndg-httpsclient==0.5.1
  - pyasn1 [required: >=0.1.1, installed: 0.4.4]
  - PyOpenSSL [required: Any, installed: 18.0.0]
    - cryptography [required: >=2.2.1, installed: 2.3]
      - asn1crypto [required: >=0.21.0, installed: 0.24.0]
      - cffi [required: >=1.7,!=1.11.3, installed: 1.11.5]
        - pycparser [required: Any, installed: 2.18]
      - enum34 [required: Any, installed: 1.1.6]
      - idna [required: >=2.1, installed: 2.7]
      - ipaddress [required: Any, installed: 1.0.22]
      - six [required: >=1.4.1, installed: 1.11.0]
    - six [required: >=1.5.2, installed: 1.11.0]
pycodestyle==2.4.0
yapf==0.22.0
```

--- a/tests/test_dynamodb2/test_dynamodb.py
+++ b/tests/test_dynamodb2/test_dynamodb.py
@@ -4946,3 +4946,86 @@ def test_multiple_updates():
         "id": {"S": "1"},
     }
     assert result == expected_result
+
+
+@mock_dynamodb2
+def test_update_item_atomic_counter():
+    table = "table_t"
+    ddb_mock = boto3.client("dynamodb", region_name="eu-west-3")
+    ddb_mock.create_table(
+        TableName=table,
+        KeySchema=[{"AttributeName": "t_id", "KeyType": "HASH"}],
+        AttributeDefinitions=[{"AttributeName": "t_id", "AttributeType": "S"}],
+        BillingMode="PAY_PER_REQUEST",
+    )
+
+    key = {"t_id": {"S": "item1"}}
+
+    ddb_mock.put_item(
+        TableName=table,
+        Item={"t_id": {"S": "item1"}, "n_i": {"N": "5"}, "n_f": {"N": "5.3"}},
+    )
+
+    ddb_mock.update_item(
+        TableName=table,
+        Key=key,
+        UpdateExpression="set n_i = n_i + :inc1, n_f = n_f + :inc2",
+        ExpressionAttributeValues={":inc1": {"N": "1.2"}, ":inc2": {"N": "0.05"}},
+    )
+    updated_item = ddb_mock.get_item(TableName=table, Key=key)["Item"]
+    updated_item["n_i"]["N"].should.equal("6.2")
+    updated_item["n_f"]["N"].should.equal("5.35")
+
+
+@mock_dynamodb2
+def test_update_item_atomic_counter_return_values():
+    table = "table_t"
+    ddb_mock = boto3.client("dynamodb", region_name="eu-west-3")
+    ddb_mock.create_table(
+        TableName=table,
+        KeySchema=[{"AttributeName": "t_id", "KeyType": "HASH"}],
+        AttributeDefinitions=[{"AttributeName": "t_id", "AttributeType": "S"}],
+        BillingMode="PAY_PER_REQUEST",
+    )
+
+    key = {"t_id": {"S": "item1"}}
+
+    ddb_mock.put_item(TableName=table, Item={"t_id": {"S": "item1"}, "v": {"N": "5"}})
+
+    response = ddb_mock.update_item(
+        TableName=table,
+        Key=key,
+        UpdateExpression="set v = v + :inc",
+        ExpressionAttributeValues={":inc": {"N": "1"}},
+        ReturnValues="UPDATED_OLD",
+    )
+    assert (
+        "v" in response["Attributes"]
+    ), "v has been updated, and should be returned here"
+    response["Attributes"]["v"]["N"].should.equal("5")
+
+    # second update
+    response = ddb_mock.update_item(
+        TableName=table,
+        Key=key,
+        UpdateExpression="set v = v + :inc",
+        ExpressionAttributeValues={":inc": {"N": "1"}},
+        ReturnValues="UPDATED_OLD",
+    )
+    assert (
+        "v" in response["Attributes"]
+    ), "v has been updated, and should be returned here"
+    response["Attributes"]["v"]["N"].should.equal("6")
+
+    # third update
+    response = ddb_mock.update_item(
+        TableName=table,
+        Key=key,
+        UpdateExpression="set v = v + :inc",
+        ExpressionAttributeValues={":inc": {"N": "1"}},
+        ReturnValues="UPDATED_NEW",
+    )
+    assert (
+        "v" in response["Attributes"]
+    ), "v has been updated, and should be returned here"
+    response["Attributes"]["v"]["N"].should.equal("8")

